### ISSUE TITLE
Teach all apply variants to deal with conflicts

### ISSUE THREAD

**Issue by greenrd** [magit/magit#2550]
I just tried to do something that had worked before: stage the region (that worked) and then reverse it (that didn't). After pressing 'v' on the staged file, or on the staged hunk (I tried it both ways), no changes were made. All the added text is still there and all the removed text is still not there, both in the buffer and in the actual file.

**Comment by npostavs** [comment]
Works for me (`Magit 2.5.0-11-g9606531, Git 2.5.0.windows.1, Emacs 24.5.1`, Windows 7). Please post more details about which versions of Emacs, magit, git, and OS you are using. And check that the problem occurs starting with `emacs -Q`.

**Comment by tarsius** [comment]
There probably is a conflict.

I would have expected it to work when using a prefix argument (which causes the use of `--3way`). Apparently not. Someone should look into that. But I don't think this is a regression - it just doesn't work.

While investigating this I also noticed that the `git apply ...` call is not logged due to the change in e62aff3fb3e207ade24ebbc879d748bf6a2ed708. @npostavs could you please figure out how to best restore the logging.

**Comment by npostavs** [comment]
> While investigating this I also noticed that the git apply ... call is not logged due to the change in e62aff3. @npostavs could you please figure out how to best restore the logging.

Ah, oops. Fixed in 83a87a4d22bb5b099f481531b0cd44053b348477. Also considering 88f4c958050c1d66b8a3b87090856ed700a8bcea which reduces the code duplication a bit, but I'm not sure if the resulting calling convention for `magit-call-process` is a good idea.

**Comment by greenrd** [comment]
Yes, it was a conflict I think - I introduced new lines nearby (before the change to be reversed).

**Comment by npostavs** [comment]
If there is a conflict, how do you see the staged hunk? I just get a non-expandable "unmerged changes" section under "Staged changes" for conflicted files. And after resolving the conflict, I can reverse changes without any problem.

**Comment by greenrd** [comment]
I wasn't doing a merge, so it wasn't a conflict resulting from a merge. I meant a conflict in that the patch wouldn't have applied (in reverse) cleanly, because the context was different.

**Comment by npostavs** [comment]
> I wasn't doing a merge, so it wasn't a conflict resulting from a merge. I meant a conflict in that the patch wouldn't have applied (in reverse) cleanly, because the context was different.

Ah, I get it now.

---

> I would have expected it to work when using a prefix argument (which causes the use of --3way). Apparently not.

I think it's because we don't give the "blob identity":

>--3way
>
>    When the patch does not apply cleanly, fall back on 3-way merge if the patch records the identity of blobs it is supposed to apply to,...

**Comment by npostavs** [comment]
> I think it's because we don't give the "blob identity", EDIT: but actually, it seems like we do have the index 4f284da..345df8a 100644 line at the top, isn't that the identity?

Oh, it seems git does see the identity, but is still not satisfied with it. Without `--3way`:
```
error: patch failed: README.md:1
error: README.md: patch does not apply
```
With `--3way`:
```
error: README.md: does not match index
```

**Comment by npostavs** [comment]
So the problem is that for the patch to apply cleanly we need to diff against the working copy, but then the actual change won't show up as a diff, because it's present in both working copy and index.

Possible workarounds:
- reduce diff context to 0
- go to `diff worktree` (`d w`) and reverse the change from there

These don't look like things that magit could do automatically though.

---

Things that do NOT work:
- `--3way`: git always refuses to apply changes when the working copy is dirty (even if the changes are far apart)
- stashing working copy changes, reversing, and then popping stash: git refuses to pop the stash when the working copy is dirty.

**Comment by tarsius** [comment]
> git always refuses to apply changes when the working copy is dirty (even if the changes are far apart)

That doesn't appear to be the case:

* Make two separate changes, stage one.

```
Unstaged changes (1)
modified   lisp/magit-apply.el
@@ -7,7 +7,7 @@
 
 ;; Author: Jonas Bernoulli <jonas@bernoul.li>
 ;; Maintainer: Jonas Bernoulli <jonas@bernoul.li>
-
+;; unstaged
 ;; Magit is free software; you can redistribute it and/or modify it
 ;; under the terms of the GNU General Public License as published by
 ;; the Free Software Foundation; either version 3, or (at your option)

Staged changes (1)
modified   lisp/magit-apply.el
@@ -1,5 +1,5 @@
 ;;; magit-apply.el --- apply Git diffs  -*- lexical-binding: t -*-
-
+;; staged
 ;; Copyright (C) 2010-2016  The Magit Project Contributors
 ;;
 ;; You should have received a copy of the AUTHORS.md file which
```

* Reverse "staged" by pressing `v` on the hunk or file.

```
Unstaged changes (1)
modified   lisp/magit-apply.el
@@ -1,5 +1,5 @@
 ;;; magit-apply.el --- apply Git diffs  -*- lexical-binding: t -*-
-;; staged
+
 ;; Copyright (C) 2010-2016  The Magit Project Contributors
 ;;
 ;; You should have received a copy of the AUTHORS.md file which
@@ -7,7 +7,7 @@
 
 ;; Author: Jonas Bernoulli <jonas@bernoul.li>
 ;; Maintainer: Jonas Bernoulli <jonas@bernoul.li>
-
+;; unstaged
 ;; Magit is free software; you can redistribute it and/or modify it
 ;; under the terms of the GNU General Public License as published by
 ;; the Free Software Foundation; either version 3, or (at your option)

Staged changes (1)
modified   lisp/magit-apply.el
@@ -1,5 +1,5 @@
 ;;; magit-apply.el --- apply Git diffs  -*- lexical-binding: t -*-
-
+;; staged
 ;; Copyright (C) 2010-2016  The Magit Project Contributors
 ;;
 ;; You should have received a copy of the AUTHORS.md file which
```

**Comment by tarsius** [comment]
After reading the documentation again and experimenting some more, I have concluded that `git apply` simply does not support applying conflicting changes to the worktree. What we would like to do is to insert conflict markers into the files, just like when there are conflicts when applying commits, i.e. when merging.

The next best thing is to use the `--reject` argument. This won't add conflict markers, but it does create "conflict files" which contain the parts of the patch that could not be applied. It also causes the non-conflicting changes to be applied.

In #2647 I start using `--reject` and stop using `--3way`.

**Comment by npostavs** [comment]
>> git always refuses to apply changes when the working copy is dirty (even if the changes are far apart)

> That doesn't appear to be the case:
[...]
> Reverse "staged" by pressing v on the hunk or file.

By "git always refuses", I meant "with `--3way` git always refuses", i.e. reversing with `--3way` by pressing `C-u v` gives 
```
1 C:/Git-2.x/mingw64/libexec/git-core/git … apply --reverse --3way -p0 --ignore-space-change -
error: lisp/magit-apply.el: does not match index
```

---

> The next best thing is to use the `--reject` argument.

Yes, seems so.

**Comment by tarsius** [comment]
>> The next best thing is to use the --reject argument.

> Yes, seems so.

But we should probably add an option for that, and a prefix argument too.

> is `--3way` really never useful

I now think it should always be used by `magit-apply` as in `(interactive (list "--3way"))`.

**Comment by tarsius** [comment]
If also experimented with creating commits, initiating a merge, and then doing a soft reset and manually removing the files inside .git which tell git that a merge is in progress. But that's pretty gross, expensive, and likely to blow up, so I don't want to go down that path.

**Comment by tarsius** [comment]
I've merged #2647.
